### ISSUE TITLE
Build shared artifact registries and HF-backed evaluation pipelines

### ISSUE THREAD

**Issue by gonzalobenegas** [Open-Athena/marin-dna#503]
## TL;DR

Add a Git-owned development registry for MarinDNA models, training datasets, and exact evaluation-dataset views.
Store reusable development artifacts and maintained pipeline results in one public-readable, write-restricted Hugging Face Storage Bucket owned by `open-athena`.
Introduce an engine-neutral `pipelines/` tree, replace the mixed `evals_v2` workflow with separate scientific pipelines, and keep filesystem-based predictions and metrics as the evaluation source of truth.

## Problem

Model and dataset metadata is currently repeated across evaluation configuration and the dashboard.
Checkpoints are commonly sourced from Marin training storage in GCS, while maintained evaluation outputs are written to S3.
The current arrangement makes it difficult to discover canonical model IDs, exact checkpoints, context windows, training-dataset lineage, evaluation-dataset views, and existing computed results.
It also encourages a new evaluation to copy model definitions or create another output namespace.

`snakemake/analysis/evals_v2` combines variant-effect prediction, nucleotide-dependency analysis, embedding UMAP, and related outputs whose inputs and execution contracts can diverge.
Future evaluations such as DART-Eval tasks may share neither VEP inference nor VEP analysis.
The workflow path also exposes Snakemake as the permanent owner even though a future pipeline may use plain Python, Iris, Metaflow, Modal, or another orchestrator.

The desired storage is open for anonymous reads and usable from EC2, Lambda, CoreWeave, Modal, and other GPU platforms.
MarinDNA should not keep paying S3 egress for public consumers.
The Hugging Face Storage Bucket pilot must verify the actual storage limits, public-read behavior, and egress terms before substantial data moves.

Hugging Face model and dataset repositories are useful publication surfaces but impose Git-oriented versioning on development artifacts that change frequently.
Hugging Face evaluation-results metadata is too restrictive for the current score layer, which needs standard errors, subsets, pipeline-specific products, and results for entities that are not necessarily Hugging Face models.

## Registry design

The development registry lives in Git:

```text
registry/
  models/<id>/README.md
  training_datasets/<id>/README.md
  evaluation_datasets/<id>/README.md
```

Each artifact or logical dataset view has one README with YAML frontmatter and free text.
Git provides review, history, and cross-reference validation.
The README remains in Git because several logical evaluation-dataset views may point to one physical bucket prefix.

Every entry includes:

- `id`, equal to its directory name and unique within its kind;
- `kind`, equal to the registry subtree;
- `description`;
- `uri`, expressed as a canonical `hf://buckets/open-athena/marin-dna/...` URI;
- `generated_by`, containing one full commit-pinned permalink to the code or pipeline that generated the artifact; and
- optional `github_issue`, containing one issue URL.

CI performs offline validation of schemas, required fields, unique IDs, directory and kind agreement, commit-pinned generation links, and registry cross-references.
Remote existence and access checks run during upload or registration because routine pull requests should not depend on Hugging Face availability or credentials.

Only development artifacts are registered.
There is no separate registry for publication mirrors.
Selected release models and datasets may be copied to ordinary Hugging Face Model or Dataset repositories and grouped in a release collection for papers or blog posts.
Those repositories are a public UI and discovery layer, while the Git development registry remains canonical for work in this repository.

## Model identity and lifecycle

The model registry contains only MarinDNA models.
Each exact checkpoint selected for evaluation or reuse receives its own registry entry and immutable bucket URI.
The entry records `context_window_bp` and a list of training-dataset IDs.
It does not record evaluation datasets or evaluation scope.

`training_run` and `training_step` record immutable provenance facts.
The registry does not classify checkpoints as final or intermediate because a checkpoint can be useful for development before a later publication decision.
Optional metadata may include actual `training_tokens`, `parameter_count`, architecture, and dashboard display fields.
Mixture weights are excluded because training mixtures may change during a run and are not reliably available for every model.

Each checkpoint directory bundles its tokenizer.
The registry has no separate tokenizer metadata.
The model record does not require its former GCS path; the commit-pinned generator and training provenance provide the durable link back to the training code.

Marin checkpoint paths preserve the original training run name and exact step:

```text
hf://buckets/open-athena/marin-dna/marin_checkpoints/<original-marin-run-name>/hf/step-<N>/
```

Example:

```text
hf://buckets/open-athena/marin-dna/marin_checkpoints/dna-bolinas-exp166-v0.1-p1B-c127da/hf/step-27329/
```

The registry ID may be cleaner than the physical path.
Uploading or registering a checkpoint must fail if the exact destination prefix already exists, with no force-overwrite mode.

## Training-dataset identity

A training-dataset entry describes one complete prepared, untokenized artifact.
It does not describe a split or configuration view.
It records generated-once `num_examples` and `num_base_pairs` statistics.
`num_base_pairs` is the sum of the stored, unpadded nucleotide sequence lengths.

The architecture does not own tokenized training data.
Marin continues to create and manage its tokenization cache in GCS.
Models refer to registered untokenized training datasets, and the checkpoint carries the tokenizer needed to reproduce tokenization.

Training-dataset builders write versioned artifacts below their owning pipeline's `results/` directory.
Once registered, an artifact directory is immutable.
A rebuild creates a new artifact directory and registry ID.

## Evaluation-dataset identity

An evaluation-dataset entry describes one exact logical view of a physical dataset.
It records the underlying bucket URI, one required `split`, optional `configuration`, `usage: development|held_out`, and generated-once `num_examples`.
It does not record `num_base_pairs`.

Several entries may share a physical URI while selecting different configurations or splits.
IDs include a human-readable artifact version and view, such as `mendelian_traits_v3_train` and `mendelian_traits_v3_test`.
A rebuilt dataset uses `v4` or another meaningful version label rather than a commit hash in the ID.

Evaluation-dataset preparation is occasional provenance work and is not maintained as a pipeline on `main`.
Preparation code stays on a permanent issue branch, and its artifacts use `issues/<issue-number>/<artifact-slug>/` storage.
Once an exact artifact or view is accepted into the Git registry, its bucket prefix is immutable and its `generated_by` field points to the commit-pinned preparation code.

Evaluation jobs pair a model ID with an exact evaluation-dataset ID.
They do not repeat split or configuration and do not synthesize a separate job ID.
Held-out entries remain unavailable to the routine development path and require the protected opt-in mechanism.

## Storage ownership and lifecycle

Use one public-readable, write-restricted bucket:

```text
hf://buckets/open-athena/marin-dna/
```

The canonical registry and user-facing identity always uses the native `hf://buckets/...` form.
Snakemake and other tools may access the same objects through `s3://marin-dna/...` with endpoint `https://s3.hf.co/open-athena`, region `us-east-1`, path-style addressing, and the Hugging Face checksum client settings.
The S3 spelling is an execution adapter and must not become artifact identity.

The bucket root has three owners:

```text
marin_checkpoints/
pipelines/
issues/
```

Registered models and datasets are immutable.
Maintained evaluation outputs are mutable canonical views.
Rerunning the same model-dataset pair replaces its canonical result only after successful completion.
Maintained results do not use commit-keyed or issue-specific score namespaces.

One-off issue artifacts use:

```text
hf://buckets/open-athena/marin-dna/issues/<issue-number>/<artifact-slug>/...
```

They are mutable while exploratory and uncited.
A prefix becomes immutable after it is cited as evidence, and changed results use a new descriptive sibling prefix.
Reusable work may later move into a maintained pipeline, while obsolete uncited artifacts may be deleted.

Ordinary development outputs do not require a manually maintained `manifest.json`.
A checksum manifest remains an optional publication or frozen-snapshot feature when it provides concrete value.

## Repository and pipeline ownership

Introduce this top-level tree additively:

```text
pipelines/
  training_datasets/<pipeline>/
  evaluations/<pipeline>/
```

The path describes artifact ownership and remains valid if a project changes orchestration engines.
There is no generic `analysis/` directory.
Interpretation workflows are evaluations and live below `pipelines/evaluations/`.
Evaluation-dataset preparation does not have a maintained category in this tree.
It remains issue-scoped work because new evaluation datasets are created infrequently.

The bucket mirrors each repository owner and appends `results/`:

```text
hf://buckets/open-athena/marin-dna/pipelines/training_datasets/<pipeline>/results/...
hf://buckets/open-athena/marin-dna/pipelines/evaluations/<pipeline>/results/...
```

`results/` is the only durable HF-backed directory inside a maintained pipeline.
Code, configuration, workflow definitions, tests, and the pipeline README remain in Git.
Logs, caches, temporary files, and orchestration metadata remain local or ephemeral.
Only the owner prefix is globally standardized; each pipeline defines its internal result schema.

Distinct scientific evaluations have separate pipelines when their inputs, inference contract, or outputs differ.
Expected owners include:

```text
pipelines/evaluations/variant_effect_prediction/
pipelines/evaluations/nucleotide_dependency/
pipelines/evaluations/embedding_umap/
pipelines/evaluations/dart_eval/
```

Linear probes remain part of VEP.
Stable dependency-light metrics may move into a shared evaluation package after at least two pipelines use the same contract.
Filtering, held-out access, applicability logic, output assembly, and heavy dependencies remain pipeline-local.
Evaluation-specific code does not move into the lightweight root `src/marin_dna/` package under the current repository boundaries.

Competitor and external baselines are out of scope.
They will not be maintained on `main` or receive architecture sub-issues.
A future one-off baseline belongs on an issue branch with issue-owned bucket storage.

## VEP configuration and outputs

Create `pipelines/evaluations/variant_effect_prediction/` additively and keep its initial implementation in Snakemake.
Leave `snakemake/analysis/evals_v2` unchanged until the new pipeline reaches output parity.

The VEP configuration contains:

- dataset execution and score settings keyed by exact evaluation-dataset IDs;
- model execution settings keyed by exact model IDs;
- global analysis parameters; and
- an explicit `evaluations` list containing `{model, dataset}` pairs.

There are no inherited or default model and dataset configurations.
Related views such as train and test may repeat settings explicitly.
The pipeline does not create a Cartesian product, accept lists inside one pair entry, configure analyses per pair, or expose a separate public-target selector.
Duplicate model-dataset pairs are invalid.

The canonical `all` target computes every standard output applicable to every explicit pair.
Users may target individual products for debugging or recovery.
VEP-specific settings such as LLR versus absolute LLR remain in the pipeline configuration because they describe scoring, not reusable dataset identity.

Jobs stage a registered checkpoint to compute-local storage and load it with `AutoModel.from_pretrained()`.
Execution groups should reuse one staged checkpoint across datasets where the compute platform permits it.
The evaluation pipeline does not write another durable checkpoint copy below its result prefix.

VEP uses this initial result layout:

```text
hf://buckets/open-athena/marin-dna/pipelines/evaluations/variant_effect_prediction/results/
  model=<model-id>/
    dataset=<evaluation-dataset-id>/
      predictions/
      metrics/
      probes/
```

## Publication and dashboard

Filesystem-based results remain canonical because they can represent predictions, standard errors, subsets, probes, and pipeline-specific outputs without forcing them into Hugging Face evaluation-result metadata.
The dashboard may read the Git registry and these result summaries directly.
The existing dashboard model configuration should seed the model registry and then be replaced by registry reads to avoid a second model catalog.

Optional Hugging Face Model and Dataset repositories remain release mirrors.
There is no requirement to publish every development checkpoint, training dataset, evaluation view, prediction, or metric as a Git-backed Hugging Face repository.

## Migration and scope

Build the new paths additively.
Current `snakemake/` workflows and S3 outputs remain readable until their replacements reach parity.
Existing S3-to-HF migration is a planned follow-up and does not block the registry or new pipeline work.
Issue #207 has been closed as not planned because it described an obsolete S3-to-S3 bucket rename.
A new migration issue will inventory retained objects, move them to the new owner paths, update active writers and readers, and request explicit approval before deleting S3 data.

This epic does not include issue #473 experiment configuration, execution, results, or analysis.
Architecture pull requests must remain isolated from that work.

Sub-issues:

1. #504: Add artifact registry schemas and offline validation.
2. #505: Validate HF Storage Bucket access for workflows.
3. #506: Register initial MarinDNA models and datasets.
4. #507: Introduce engine-neutral pipeline and bucket ownership.
5. #508: Add an HF-backed variant-effect prediction pipeline.
6. #509: Split maintained non-VEP workflows out of `evals_v2`.
7. #510: Read artifact registries and HF results in the dashboard.
8. #511: Migrate retained S3 artifacts to HF and retire `oa-bolinas`.

Done when:

- All sub-issues are complete or explicitly removed from scope in this body.
- Registry schemas, offline validation, and representative real entries are merged.
- The HF Storage Bucket access path is validated from a workflow and non-local compute.
- Maintained pipeline outputs follow the engine-neutral ownership convention.
- VEP and the retained non-VEP evaluations reach parity in independent pipelines.
- The dashboard no longer maintains a duplicate model and dataset catalog.
- Retained S3 artifacts have moved to HF, active readers and writers have cut over, and S3 retirement has received separate approval.

**Comment by gonzalobenegas** [comment]
About the HF bucket, I still want to understand how this would work with a large HAL file. Maybe it would be a lot of egress costs.

Perhaps we could make the HAL file smaller by filtering species (e.g. one per family) or transforming it into another format, such as maf or taf (see https://github.com/ComparativeGenomicsToolkit/taffy).
Also remove dups.
